### PR TITLE
Introduce "cp" command

### PR DESCRIPTION
**Pull request by ndeloof**
to mimic `docker cp` UX but using service abstraction
delegates to the actual docker cp command once target container has been
resolved

Mostly to check interest by maintainers, or maybe we should just reject #5523 and document workaround.

### CODE REVIEW

File: compose/cli/main.py [hunk 1]
```diff
@@ -757,6 +761,44 @@ def ps(self, options):
...
+        src_service, src_path = splitCpArg(source)
+        dest_service, dest_path = splitCpArg(destination)
+        if src_service is not None and dest_service is not None:
+            raise UserError('copying between containers is not supported')
+        service = src_service or dest_service
+        if service is None:
+            raise UserError('neither source nor destination path do refer to a service')
+        containers = self.project.containers(service_names=[service], stopped=True,
+                                             one_off=OneOffFilter.include)
+        if len(containers) == 0:
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
        if not containers:
```

File: compose/cli/main.py [hunk 2]
```diff
@@ -757,6 +761,44 @@ def ps(self, options):
...
+        if len(containers) > 1:
+            raise UserError('multiple containers match the service definition')
+        container = containers[0]
+
+        args = ["cp"]
+        if options["--archive"]:
+            args += ["--archive"]
+        if options["--follow-link"]:
+            args += ["--follow-link"]
+        args += [src_path if src_service is None else container.id+":"+src_path]
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
        args += [src_path if not src_service else container.id+":"+src_path]
```

File: compose/cli/main.py [hunk 3]
```diff
@@ -757,6 +761,44 @@ def ps(self, options):
...
+        args += [dest_path if dest_service is None else container.id + ":" + dest_path]
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
        args += [dest_path if not dest_service else container.id + ":" + dest_path]
```

File: compose/cli/main.py [hunk 4]
```diff
@@ -757,6 +761,44 @@ def ps(self, options):
...
+        """
+        source = options['SOURCE']
+        destination = options['DESTINATION']
+
+        src_service, src_path = splitCpArg(source)
+        dest_service, dest_path = splitCpArg(destination)
+        if src_service is not None and dest_service is not None:
+            raise UserError('copying between containers is not supported')
+        service = src_service or dest_service
+        if service is None:
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
        if not service:
```

File: compose/cli/main.py [hunk 5]
```diff
@@ -757,6 +761,44 @@ def ps(self, options):
...
+        Options:
+            -a, --archive       Archive mode (copy all uid/gid information)
+            -L, --follow-link   Always follow symbol link in SRC_PATH
+        """
+        source = options['SOURCE']
+        destination = options['DESTINATION']
+
+        src_service, src_path = splitCpArg(source)
+        dest_service, dest_path = splitCpArg(destination)
+        if src_service is not None and dest_service is not None:
```
Comment (ulyssessouza) [inline compose/cli/main.py]:
```suggestion
        if src_service and dest_service:
```

File: compose/cli/main.py [hunk 6]
```diff
@@ -710,7 +712,9 @@ def ps(self, options):
         Options:
             -q, --quiet          Only display IDs
             --services           Display services
-            --filter KEY=VAL     Filter services by a property
+            --filter KEY=VAL     Filter s        # TODO add support for "-" as path
+
+ervices by a property
```
Comment (ulyssessouza) [inline compose/cli/main.py]: Didn't understand this part... Is it intentional?

Comment (ndeloof) [reply compose/cli/main.py]: bad copy/paste

File: compose/cli/utils.py [hunk 7]
```diff
@@ -153,3 +153,12 @@ def binarystr_to_unicode(s):
...
     return s.decode('utf-8', 'replace')
+
+
+# See cli/command/container/cp.go#splitCpArg
+def splitCpArg(path):
+    if not os.path.isabs(path) and not path.startswith('.'):
+        parts = path.split(":", 2)
+        if len(parts) == 2:
+            return (parts[0], parts[1])
+    return (None, path)
```
Comment (joseangel-sc) [inline compose/cli/utils.py]:
```suggestion
    return None, path
```

File: compose/cli/utils.py [hunk 8]
```diff
@@ -153,3 +153,12 @@ def binarystr_to_unicode(s):
...
             pass
     return s.decode('utf-8', 'replace')
+
+
+# See cli/command/container/cp.go#splitCpArg
+def splitCpArg(path):
+    if not os.path.isabs(path) and not path.startswith('.'):
+        parts = path.split(":", 2)
+        if len(parts) == 2:
+            return (parts[0], parts[1])
```
Comment (joseangel-sc) [inline compose/cli/utils.py]:
```suggestion
            return parts[0], parts[1]
```

File: compose/cli/main.py [hunk 9]
```diff
@@ -774,14 +774,14 @@ def cp(self, options):
 
         src_service, src_path = splitCpArg(source)
         dest_service, dest_path = splitCpArg(destination)
-        if src_service is not None and dest_service is not None:
+        if src_service and dest_service:
             raise UserError('copying between containers is not supported')
         service = src_service or dest_service
-        if service is None:
+        if not service:
```
Comment (joseangel-sc) [inline compose/cli/main.py]: I don't think this is the exact same behavior and could lead to unwanted errors

Comment (ndeloof) [reply compose/cli/main.py]: According to https://stackoverflow.com/questions/3901144/difference-between-if-x-and-if-x-is-not-none I think we are safe here

File: compose/cli/utils.py (hunk 7 above)
Comment (ndeloof) [reply compose/cli/utils.py]: :+1:

**Comment by barelnir** [comment]
guys,

what is the status of this pull request? can it be merged to master as part of next docker-compose version?

**Comment by ndeloof** [comment]
This is just a proposal.
- miss test case to cover this new feature
- miss documentation
- need to confirm with other maitainers delegating to `docker cp` is the right thing to do, or if we want to rely on docker-py (more code, more bugs imho)

**Pull request closed without merging**